Library/Homebrew/upgrade.rb

```diff
@@ -276,6 +276,20 @@ def check_installed_dependents(
         installed_formulae.flat_map(&:runtime_installed_formula_dependents)
                           .uniq
                           .select(&:outdated?)
+
+      # Ensure we never attempt a source build for outdated dependents of upgraded formulae.
+      outdated_dependents, skipped_dependents = outdated_dependents.partition do |dependent|
+        dependent.bottled? && dependent.deps.all?(&:bottled?)
```

Comment:
Does this work in all the situations we expect e.g. earlier OS version bottles, `:all` bottles etc.?

Reply:
Yes. `#bottled?` returns `true` for `:all` bottles and earlier OS version bottles, and also returns `false` for non-relocatable bottles in a non-default prefix.